File: Old Java Solutions/Dynamic Programming/Merge Intervals.java
class Solution {


    public int[][] merge(int[][] intervals) {
        
        if (intervals.length <= 1) {
            return intervals;
        }

        List<int[]> result = new ArrayList<>();

        // Sort based on first number
        Arrays.sort(intervals, (arr1, arr2) -> Integer.compare(arr1[0], arr2[0]));

        result.add(intervals[0]);

        // Tracks the current interval, which will be updated to the interval with the highest
        // starting number when encountered
        int[] currentInterval = intervals[0];
        for (int[] interval : intervals) {

            // merge
            if (interval[0] <= currentInterval[1]) {
                currentInterval[1] = Math.max(currentInterval[1], interval[1]);
            } else {

                // Add new interval, will be updated in result if there are future merges
                result.add(interval);

                currentInterval = interval;
            }

        }

        int[][] resultArray = new int[result.size()][2];

        for (int i = 0; i < result.size(); i ++) {
            resultArray[i] = result.get(i);
        }

        return resultArray;




    }
}

File: Interval/Merge Intervals.py
class Solution:

    def can_merge(self, interval1, interval2):

        return interval2[0] <= interval1[1]

    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        
        if len(intervals) == 0:
            return []

        if len(intervals) == 1:
            return intervals

        # sort by starting time of each interval
        intervals.sort(key=lambda x: x[0])

        result = []

        current_merged_interval = intervals[0]

        for i in range(1, len(intervals)):

            current_interval = intervals[i]

            if self.can_merge(current_merged_interval, current_interval):

                current_merged_interval = [current_merged_interval[0], max(current_merged_interval[1], current_interval[1])]

            else:
                result.append(current_merged_interval)
                current_merged_interval = current_interval


        result.append(current_merged_interval)


        return result





